x/fees/keeper/msg_server.go

```diff
@@ -19,65 +20,95 @@ func (k Keeper) RegisterFee(
-		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "deployer account not found %s", msg.DeployerAddress)
+	deployerAcc := k.evmKeeper.GetAccountWithoutBalance(ctx, common.BytesToAddress(deployer))
+	if deployerAcc == nil {
+		return nil, sdkerrors.Wrapf(
+			sdkerrors.ErrNotFound,
+			"deployer account not found %s", msg.DeployerAddress,
+		)
 	}
-	if deployerAccount.IsContract() {
-		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "deployer cannot be a contract %s", msg.DeployerAddress)
+
+	if deployerAcc.IsContract() {
+		return nil, sdkerrors.Wrapf(
+			types.ErrFeesDeployerIsNotEOA,
+			"deployer cannot be a contract %s", msg.DeployerAddress,
+		)
 	}
 
-	var withdrawal sdk.AccAddress
-	if msg.WithdrawAddress != "" {
-		withdrawal = sdk.MustAccAddressFromBech32(msg.WithdrawAddress)
+	var withdraw sdk.AccAddress
+	if msg.WithdrawerAddress != "" && msg.WithdrawerAddress != msg.DeployerAddress {
+		withdraw = sdk.MustAccAddressFromBech32(msg.WithdrawerAddress)
 	}
 
+	derivedContractAddr := common.BytesToAddress(deployer)
+
 	// the contract can be directly deployed by an EOA or created through one
 	// or more factory contracts. If it was deployed by an EOA account, then
 	// msg.Nonces contains the EOA nonce for the deployment transaction.
 	// If it was deployed by one or more factories, msg.Nonces contains the EOA
 	// nonce for the origin factory contract, then the nonce of the factory
 	// for the creation of the next factory/contract.
-	addrDerivationCostCreate := k.GetParams(ctx).AddrDerivationCostCreate
-	derivedContractAddr := common.BytesToAddress(deployer)
 	for _, nonce := range msg.Nonces {
 		ctx.GasMeter().ConsumeGas(
-			addrDerivationCostCreate,
+			params.AddrDerivationCostCreate,
 			"fees registration: address derivation CREATE opcode",
 		)
+
 		derivedContractAddr = crypto.CreateAddress(derivedContractAddr, nonce)
 	}
 
 	if contract != derivedContractAddr {
 		return nil, sdkerrors.Wrapf(
 			sdkerrors.ErrorInvalidSigner,
-			"not contract deployer or wrong nonce: expected %s instead of %s", derivedContractAddr.String(),
-			msg.ContractAddress,
+			"not contract deployer or wrong nonce: expected %s instead of %s",
+			derivedContractAddr, msg.ContractAddress,
 		)
 	}
 
 	// contract must already be deployed, to avoid spam registrations
 	contractAccount := k.evmKeeper.GetAccountWithoutBalance(ctx, contract)
+
 	if contractAccount == nil || !contractAccount.IsContract() {
-		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "contract has no code %s", msg.ContractAddress)
+		return nil, sdkerrors.Wrapf(
+			types.ErrFeesNoContractDeployed,
+			"no contract code found at address %s", msg.ContractAddress,
+		)
+	}
+
+	// prevent storing the same address for deployer and withdrawer
+	fee := types.NewFee(contract, deployer, withdraw)
+	k.SetFee(ctx, fee)
+	k.SetDeployerMap(ctx, deployer, contract)
+
+	// NOTE: only set withdraw map if address is not empty
+
+	withdrawAddr := msg.DeployerAddress
```

Comment:
The use of `withdrawAddr` and `withdraw` here is not immediately clear. I think either of the combinations 
- `assignedWithdrawerAddress` and `withdrawerAddress`
- `withdrawerAddress` and `msgWithdrawerAddress` 

might be more suitable.

Reply:
Nice catch, I changed all variables that represented addresses to drop the `address` in the name and in case of an account added `account`. Also added an `effectiveWithdrawer` workding to differentiate it from the temporary withdrawer address.